### PR TITLE
Fix documented exception types thrown by methods in device client.

### PR DESCRIPTION
**Pull request by azabbasi**
Fixe #2246

### CODE REVIEW

File: iothub/device/tests/DeviceClientTests.cs [hunk 1]
```diff
@@ -1773,9 +1775,291 @@ public void DeviceClient_InitWithHttpTransportButNoModelId_DoesNotThrow()
                 .NotThrow();
         }
 
+
+        [Description("Testing current behavior, The logically correct expected exception should be OperationCancelledException but current expected exception is IotHubCommunicationException since we try to avoid behavior change.")]
```
Comment (azabbasi) [inline iothub/device/tests/DeviceClientTests.cs:1779]: Currently looking to refactor the common code out to reduce code.

Comment (azabbasi) [reply iothub/device/tests/DeviceClientTests.cs:1779]: too many specific variables are invovled. will leave it as is.

**Review by timtay-microsoft — APPROVED** [review]

File: iothub/device/tests/DeviceClientTests.cs [hunk 2]
```diff
@@ -1773,9 +1775,291 @@ public void DeviceClient_InitWithHttpTransportButNoModelId_DoesNotThrow()
...
+        [TestMethod]
+        [ExpectedException(typeof(IotHubCommunicationException))]
```
Comment (drwill-ms) [inline iothub/device/tests/DeviceClientTests.cs:1781]: Consider using FluentAssertions instead. You can validate a method throws.

https://fluentassertions.com/exceptions/

File: iothub/device/tests/DeviceClientTests.cs [hunk 3]
```diff
@@ -1773,9 +1775,291 @@ public void DeviceClient_InitWithHttpTransportButNoModelId_DoesNotThrow()
...
+        public async Task DeviceClient_ReceiveAsync_Cancelled_ThrowsCorrectException()
+        {
+            using var deviceClient = DeviceClient.CreateFromConnectionString(FakeConnectionString);
```
Comment (drwill-ms) [inline iothub/device/tests/DeviceClientTests.cs:1784]: // arrange

// act

// assert

?

Comment (azabbasi) [reply iothub/device/tests/DeviceClientTests.cs:1784]: sure, I can add that.

File: iothub/device/tests/DeviceClientTests.cs [hunk 4]
```diff
@@ -1773,9 +1775,291 @@ public void DeviceClient_InitWithHttpTransportButNoModelId_DoesNotThrow()
                 .NotThrow();
         }
 
+
+        [Description("Testing current behavior, The logically correct expected exception should be OperationCancelledException but current expected exception is IotHubCommunicationException since we try to avoid behavior change.")]
+        [TestMethod]
+        [ExpectedException(typeof(IotHubCommunicationException))]
+        public async Task DeviceClient_ReceiveAsync_Cancelled_ThrowsCorrectException()
```
Comment (drwill-ms) [inline iothub/device/tests/DeviceClientTests.cs:1782]: What do you think about the method name illustrating that we are verifying and preserving legacy behavior?

File: iothub/device/tests/DeviceClientTests.cs [hunk 5]
```diff
@@ -1773,9 +1775,291 @@ public void DeviceClient_InitWithHttpTransportButNoModelId_DoesNotThrow()
...
+            using var cts = new CancellationTokenSource();
+            cts.Cancel();
+
+            await deviceClient.OpenAsync(cts.Token);
+        }
+
+        [Description("Testing current behavior, The logically correct expected exception should be OperationCancelledException but current expected exception is IotHubCommunicationException since we try to avoid behavior change.")]
+        [TestMethod]
+        [ExpectedException(typeof(IotHubCommunicationException))]
+        public async Task DeviceClient_AbandoneAsync_Cancelled_ThrowsCorrectException()
```
Comment (drwill-ms) [inline iothub/device/tests/DeviceClientTests.cs:1918]:
```suggestion
        public async Task DeviceClient_AbandonAsync_Cancelled_ThrowsCorrectException()
```

File: iothub/device/tests/DeviceClientTests.cs [hunk 6]
```diff
@@ -1773,9 +1775,291 @@ public void DeviceClient_InitWithHttpTransportButNoModelId_DoesNotThrow()
...
+            using var cts = new CancellationTokenSource();
+            cts.Cancel();
+
+            await deviceClient.RejectAsync("SomeToken", cts.Token);
+        }
+
+        [Description("Testing current behavior, The logically correct expected exception should be OperationCancelledException but current expected exception is IotHubCommunicationException since we try to avoid behavior change.")]
+        [TestMethod]
+        [ExpectedException(typeof(IotHubCommunicationException))]
+        public async Task DeviceClient_SendAsync_Cancelled_ThrowsCorrectException()
```
Comment (drwill-ms) [inline iothub/device/tests/DeviceClientTests.cs:1863]:
```suggestion
        public async Task DeviceClient_SendEventAsync_Cancelled_ThrowsCorrectException()
```

File: iothub/device/tests/DeviceClientTests.cs [hunk 7]
```diff
@@ -1773,9 +1775,291 @@ public void DeviceClient_InitWithHttpTransportButNoModelId_DoesNotThrow()
...
+            using var cts = new CancellationTokenSource();
+            cts.Cancel();
+
+            await deviceClient.AbandonAsync("SomeLockToken", cts.Token);
+        }
+
+        [Description("Testing current behavior, The logically correct expected exception should be OperationCancelledException but current expected exception is IotHubCommunicationException since we try to avoid behavior change.")]
+        [TestMethod]
+        [ExpectedException(typeof(IotHubCommunicationException))]
+        public async Task DeviceClient_UpdateReportedProperties_Cancelled_ThrowsCorrectException()
```
Comment (drwill-ms) [inline iothub/device/tests/DeviceClientTests.cs:1945]:
```suggestion
        public async Task DeviceClient_UpdateReportedPropertiesAsync_Cancelled_ThrowsCorrectException()
```

File: iothub/device/tests/DeviceClientTests.cs [hunk 8]
```diff
@@ -1773,9 +1775,291 @@ public void DeviceClient_InitWithHttpTransportButNoModelId_DoesNotThrow()
...
+            using var cts = new CancellationTokenSource();
+            cts.Cancel();
+
+            await deviceClient.CloseAsync(cts.Token);
+        }
+
+        [Description("CloseAsync will not go through the ErrorDelegationHandler hence the exception will bubble up directly to the caller with no transformation.")]
+        [TestMethod]
+        [ExpectedException(typeof(TaskCanceledException))]
+        public async Task DeviceClient_SetDesiredPropertyCallback_Cancelled_ThrowsCorrectException()
```
Comment (drwill-ms) [inline iothub/device/tests/DeviceClientTests.cs:2026]:
```suggestion
        public async Task DeviceClient_SetDesiredPropertyCallbackAsync_Cancelled_ThrowsCorrectException()
```

File: iothub/device/tests/DeviceClientTests.cs [hunk 9]
```diff
@@ -1773,9 +1775,291 @@ public void DeviceClient_InitWithHttpTransportButNoModelId_DoesNotThrow()
...
+
+            deviceClient.InnerHandler = errorHandler;
+
+            // We will pass in an expired token to make sure the ErrorDelegationHandler or the InternalClient will not throw a different type of exception.
+            // This can happen if the ErrorDelegationHandler or the InternalClient checks the token for expiry before calling into the protocol specific delegate.
+            using var cts = new CancellationTokenSource();
+            cts.Cancel();
+
+            await deviceClient.SetDesiredPropertyUpdateCallbackAsync(
+                (patch, context) =>
```
Comment (drwill-ms) [inline iothub/device/tests/DeviceClientTests.cs:2052]: You can do a one liner: `(patch, context) => Task.FromResult(true);`

File: iothub/device/tests/DeviceClientTests.cs (hunk 4 above)
Comment (azabbasi) [reply iothub/device/tests/DeviceClientTests.cs:1782]: I like it ! 
Decided on `DeviceClient_ReceiveAsync_MaintainLegacyExecptionBehavior`

File: iothub/device/tests/DeviceClientTests.cs (hunk 2 above)
Comment (azabbasi) [reply iothub/device/tests/DeviceClientTests.cs:1781]: Fixing the whole file.

**Review by drwill-ms — APPROVED** [review]

**Pull request merged**